numpy/_core/defchararray.pyi

```diff
@@ -464,66 +468,86 @@ __all__: list[str]
 def equal(x1: U_co, x2: U_co) -> NDArray[np.bool]: ...
 @overload
 def equal(x1: S_co, x2: S_co) -> NDArray[np.bool]: ...
+@overload
+def equal(x1: T_co, x2: T_co) -> NDArray[np.bool]: ...
 
 @overload
 def not_equal(x1: U_co, x2: U_co) -> NDArray[np.bool]: ...
 @overload
 def not_equal(x1: S_co, x2: S_co) -> NDArray[np.bool]: ...
+@overload
+def not_equal(x1: T_co, x2: T_co) -> NDArray[np.bool]: ...
 
 @overload
 def greater_equal(x1: U_co, x2: U_co) -> NDArray[np.bool]: ...
 @overload
 def greater_equal(x1: S_co, x2: S_co) -> NDArray[np.bool]: ...
+@overload
+def greater_equal(x1: T_co, x2: T_co) -> NDArray[np.bool]: ...
 
 @overload
 def less_equal(x1: U_co, x2: U_co) -> NDArray[np.bool]: ...
 @overload
 def less_equal(x1: S_co, x2: S_co) -> NDArray[np.bool]: ...
+@overload
+def less_equal(x1: T_co, x2: T_co) -> NDArray[np.bool]: ...
 
 @overload
 def greater(x1: U_co, x2: U_co) -> NDArray[np.bool]: ...
 @overload
 def greater(x1: S_co, x2: S_co) -> NDArray[np.bool]: ...
+@overload
+def greater(x1: T_co, x2: T_co) -> NDArray[np.bool]: ...
 
 @overload
 def less(x1: U_co, x2: U_co) -> NDArray[np.bool]: ...
 @overload
 def less(x1: S_co, x2: S_co) -> NDArray[np.bool]: ...
+@overload
+def less(x1: T_co, x2: T_co) -> NDArray[np.bool]: ...
 
-# String operations
 @overload
-def add(x1: U_co, x2: U_co) -> NDArray[str_]: ...
+def add(x1: U_co, x2: U_co) -> NDArray[np.str_]: ...
 @overload
-def add(x1: S_co, x2: S_co) -> NDArray[bytes_]: ...
+def add(x1: S_co, x2: S_co) -> NDArray[np.bytes_]: ...
+@overload
+def add(x1: T_co, x2: T_co) -> AR_T: ...
 
 @overload
-def multiply(a: U_co, i: i_co) -> NDArray[str_]: ...
+def multiply(a: U_co, i: i_co) -> NDArray[np.str_]: ...
+@overload
+def multiply(a: S_co, i: i_co) -> NDArray[np.bytes_]: ...
 @overload
-def multiply(a: S_co, i: i_co) -> NDArray[bytes_]: ...
+def multiply(a: T_co, i: i_co) -> AR_T: ...
 
 @overload
-def mod(a: U_co, value: Any) -> NDArray[str_]: ...
+def mod(a: U_co, value: Any) -> NDArray[np.str_]: ...
 @overload
-def mod(a: S_co, value: Any) -> NDArray[bytes_]: ...
+def mod(a: S_co, value: Any) -> NDArray[np.bytes_]: ...
+@overload
+def mod(a: T_co, value: Any) -> AR_T: ...
 
 @overload
 def capitalize(a: U_co) -> NDArray[str_]: ...
 @overload
 def capitalize(a: S_co) -> NDArray[bytes_]: ...
+@overload
+def capitalize(a: T_co) -> AR_T: ...
 
 @overload
 def center(a: U_co, width: i_co, fillchar: U_co = ...) -> NDArray[str_]: ...
 @overload
 def center(a: S_co, width: i_co, fillchar: S_co = ...) -> NDArray[bytes_]: ...
+@overload
+def center(a: T_co, width: i_co, fillchar: T_co = ...) -> AR_T: ...
```

Comment:
Now that `T_co` includes more stuff, it overlaps with `U_co`. 
But these first and thirds overloads are incompatible return types.
Pyright/pylance reports this as e.g.

```
Overload 1 for "add" overlaps overload 3 and returns an incompatible type
```

This is a very common issue within the numpy stubs actually, and I've made it a couple of times myself. 
It's not really obvious to see why this is an issue, especially in this particular case.
So here's an example of what exactly can go wrong:

```python
vec_str: list[str]
vec_str_or_T: list[str] | ndarray[tuple[int], StringDType]

reveal_type(capitalize(vec_str))
reveal_type(capitalize(vec_str_or_T))
```

and the (pyright) output is

```
Type of "capitalize(vec_str)" is "ndarray[tuple[int, ...], dtype[str_]]"
Type of "capitalize(vec_str_or_T)" is "ndarray[tuple[int, ...], StringDType]"
```

The second dtype is wrong: it should've been `dtype[str_] | StringDType` instead.
So effectively, the `list[str]` part of the union was ignored.

---

To solve this case you'll need an extra overload to completely fix this:

```python
@overload
def capitalize(a: U_co) -> NDArray[str_]: ...
@overload
def capitalize(a: S_co) -> NDArray[bytes_]: ...
@overload
def capitalize(a: _SupportsArray[StringDType]) -> AR_T: ...
@overload
def capitalize(a: T_co) -> ndarray[_Shape, dtype[str_] | StringDType]: ...
```

Here, the 3rd overload handles the "only StringDType" case, and the 4th one the "mixed str and StringDType" case

This same fix can also be applied to the other functions whose overloads have incompatible return types, such as `add` and `replace`. 
So the functions that return the same type from each overload, e.g. `split`, don't have this problem.

---

In https://github.com/numpy/numpy/issues/27032#issuecomment-2278958251 I explained this "overlapping overloads" in more detail.

Reply:
BTW, you could also choose to drop the 4th overload in my proposed fix, in case that you think that the additional complexity isn't worth it.

The consequence is that passing `list[str] | ndarray[tuple[int], StringDType]` is effectively disallowed.

Reply:
I'm trying to implement this suggestion but got stuck.

For `add` I have:

```python
@overload
def add(x1: U_co, x2: U_co) -> NDArray[np.str_]: ...
@overload
def add(x1: S_co, x2: S_co) -> NDArray[np.bytes_]: ...
@overload
def add(x1: _StringDTypeSupportsArray, x2: _StringDTypeSupportsArray) -> _StringDTypeArray: ...
@overload
def add(x1: T_co, T_co) -> _StringDTypeOrUnicodeArray: ...
```

Where 

```
_StringDTypeArray: TypeAlias = np.ndarray[_Shape, np.dtypes.StringDType]
_T_SA: TypeAlias = _SupportsArray[np.dtypes.StringDType]
_StringDTypeOrUnicodeArray: TypeAlias = np.ndarray[_Shape, np.dtype[np.str_] | np.dtypes.StringDType]
```

That fixes the error you pasted above. Woohoo!

But then for `multiply`, I now have:

```python
@overload
def multiply(a: U_co, i: i_co) -> NDArray[np.str_]: ...
@overload
def multiply(a: S_co, i: i_co) -> NDArray[np.bytes_]: ...
@overload
def multiply(a: _StringDTypeSupportsArray, i: i_co) -> _StringDTypeArray: ...
@overload
def multiply(a: T_co, i: i_co) -> _StringDTypeOrUnicodeArray: ...
```

And now pyright complains:

```
defchararray.pyi:524:5 - error: Overload 2 for "multiply" overlaps overload 4 and returns an incompatible type
```

I guess this is because the second argument is integer? Any idea how to work around this?

Reply:
I don't think that the seconds argument is relevant here, because it's the same everywhere. Sometimes in cases like these it can help to distribute the dtype over arrays, i.e.

```python
_StringDTypeOrUnicodeArray: TypeAlias = np.ndarray[_Shape, np.dtype[np.str_]] | np.ndarray[_Shape, np.dtypes.StringDType]
```